File: dashboard/frontend/src/api/news.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiFetch } from './client'
import { POLL } from '@/lib/polling'
import type {
  NewsItem, NewsSource, NewsRun, NewsStats, NewsStatus, NewsDeepSearchResult,
} from '@/lib/types'

export interface NewsListResponse {
  total: number
  limit: number
  offset: number
  results: NewsItem[]
}

export interface NewsRunsResponse {
  results: NewsRun[]
}

export interface NewsListFilters {
  status?: NewsStatus
  hide_statuses?: string  // CSV
  cve?: string
  kev_listed?: boolean
  rce?: boolean
  red_team_only?: boolean
  q?: string
  since?: string  // ISO timestamp; items with last_seen >= since
  include_deleted?: boolean
  limit?: number
  offset?: number
}

function _qs(filters: NewsListFilters | undefined): string {
  if (!filters) return ''
  const p = new URLSearchParams()
  Object.entries(filters).forEach(([k, v]) => {
    if (v === undefined || v === null || v === '') return
    p.set(k, String(v))
  })
  const s = p.toString()
  return s ? `?${s}` : ''
}

export function useNewsItems(filters?: NewsListFilters) {
  return useQuery({
    queryKey: ['news', 'items', filters || {}],
    queryFn: () => apiFetch<NewsListResponse>(`/news/items${_qs(filters)}`),
    refetchInterval: POLL.NORMAL,
  })
}

export function useNewsItem(id: string | null) {
  return useQuery({
    queryKey: ['news', 'item', id],
    queryFn: () => apiFetch<NewsItem>(`/news/items/${id}`),
    enabled: !!id,
  })
}

export function useNewsStats() {
  return useQuery({
    queryKey: ['news', 'stats'],
    queryFn: () => apiFetch<NewsStats>('/news/stats'),
    refetchInterval: POLL.NORMAL,
  })
}

export function useNewsSources() {
  return useQuery({
    queryKey: ['news', 'sources'],
    queryFn: () => apiFetch<{ results: NewsSource[] }>('/news/sources'),
    refetchInterval: POLL.SLOW,
  })
}

export function useNewsRuns(limit = 20) {
  return useQuery({
    queryKey: ['news', 'runs', limit],
    queryFn: () => apiFetch<NewsRunsResponse>(`/news/runs?limit=${limit}`),
    refetchInterval: POLL.NORMAL,
  })
}

export function useNewsRun(runId: string | null) {
  return useQuery({
    queryKey: ['news', 'run', runId],
    queryFn: () => apiFetch<NewsRun>(`/news/runs/${runId}`),
    enabled: !!runId,
    refetchInterval: (data) => (data as any)?.status === 'running' ? 1500 : false,
  })
}

export function useTriggerIngest() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (sourceId?: string) =>
      apiFetch<{ ok: boolean; run_id: string }>(
        sourceId ? `/news/ingest?source_id=${sourceId}` : '/news/ingest',
        { method: 'POST' },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'runs'] })
    },
  })
}

export function useUpdateNewsItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...patch }: { id: string } & Partial<{
      status: NewsStatus; notes: string; tags: string[]; acknowledged_by: string
    }>) =>
      apiFetch<NewsItem>(`/news/items/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(patch),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
      qc.invalidateQueries({ queryKey: ['news', 'stats'] })
    },
  })
}

export function useBulkNewsAction() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: { ids: string[]; action: string; value?: string }) =>
      apiFetch<{ updated: number }>('/news/items/bulk', {
        method: 'POST',
        body: JSON.stringify(body),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
      qc.invalidateQueries({ queryKey: ['news', 'stats'] })
    },
  })
}

export function useMatchAssets() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<{ ok: boolean; asset_hits: number }>(`/news/items/${id}/match-assets`, {
        method: 'POST',
      }),
    onSuccess: (_data, id) => {
      qc.invalidateQueries({ queryKey: ['news', 'item', id] })
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
    },
  })
}

export function useGithubSearch() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<{ ok: boolean; repos: number }>(`/news/items/${id}/github-search`, {
        method: 'POST',
      }),
    onSuccess: (_data, id) => {
      qc.invalidateQueries({ queryKey: ['news', 'item', id] })
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
    },
  })
}

export function useEnrichItem() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<{ ok: boolean; enriched: number }>(`/news/items/${id}/enrich`, {
        method: 'POST',
      }),
    onSuccess: (_data, id) => {
      qc.invalidateQueries({ queryKey: ['news', 'item', id] })
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
    },
  })
}

export function useDeepSearch() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: { topic: string; refresh_llm?: boolean; max_items?: number; include_deleted?: boolean }) =>
      apiFetch<{ ok: boolean; run_id: string; topic: string }>('/news/deep-search', {
        method: 'POST',
        body: JSON.stringify(body),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'items'] })
      qc.invalidateQueries({ queryKey: ['news', 'runs'] })
    },
  })
}

export function useUpdateSource() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...patch }: { id: string } & Partial<{
      enabled: boolean; url: string; name: string
    }>) =>
      apiFetch<{ ok: boolean }>(`/news/sources/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(patch),
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['news', 'sources'] }),
  })
}

export function useRefetchSource() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<{ ok: boolean; run_id: string }>(`/news/sources/${id}/refetch`, {
        method: 'POST',
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'runs'] })
      qc.invalidateQueries({ queryKey: ['news', 'sources'] })
    },
  })
}

export function useCreateSource() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: { name: string; url: string; parser?: string; enabled?: boolean }) =>
      apiFetch<NewsSource>('/news/sources', {
        method: 'POST',
        body: JSON.stringify(body),
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['news', 'sources'] })
    },
  })
}

File: dashboard/bff/routers/news.py
"""BFF proxy for the rag-api /news endpoints — News Intelligence feature."""
from typing import Optional, Any

import httpx
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from config import get_settings
from engagement import engagement_headers
from utils import safe_json

router = APIRouter()

_TIMEOUT = httpx.Timeout(60.0, connect=10.0)


def _client() -> httpx.AsyncClient:
    return httpx.AsyncClient(verify=False, timeout=_TIMEOUT)


# --- Models ---

class NewsItemPatch(BaseModel):
    status: Optional[str] = None
    notes: Optional[str] = None
    tags: Optional[list[str]] = None
    acknowledged_by: Optional[str] = None


class NewsBulkBody(BaseModel):
    ids: list[str]
    action: str
    value: Optional[str] = None


class NewsSourcePatch(BaseModel):
    enabled: Optional[bool] = None
    url: Optional[str] = None
    name: Optional[str] = None


class NewsDeepSearchBody(BaseModel):
    topic: str
    include_deleted: Optional[bool] = False
    refresh_llm: Optional[bool] = False
    max_items: Optional[int] = 50


def _h() -> dict:
    s = get_settings()
    return {"x-api-key": s.api_key, **engagement_headers()}


def _u(path: str) -> str:
    return f"{get_settings().rag_api_url}{path}"


# --- Routes ---

@router.post("/api/news/ingest")
async def news_ingest(source_id: Optional[str] = Query(None)):
    params = {"source_id": source_id} if source_id else {}
    async with _client() as c:
        resp = await c.post(_u("/news/ingest"), params=params, headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/runs/{run_id}")
async def news_run_status(run_id: str):
    async with _client() as c:
        resp = await c.get(_u(f"/news/runs/{run_id}"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/runs")
async def news_runs_list(limit: int = Query(20, ge=1, le=200)):
    async with _client() as c:
        resp = await c.get(_u("/news/runs"), params={"limit": limit}, headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/items")
async def news_items_list(
    status: Optional[str] = Query(None),
    hide_statuses: Optional[str] = Query(None),
    cve: Optional[str] = Query(None),
    kev_listed: Optional[bool] = Query(None),
    rce: Optional[bool] = Query(None),
    red_team_only: bool = Query(False),
    q: Optional[str] = Query(None),
    since: Optional[str] = Query(None),
    include_deleted: bool = Query(False),
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
):
    params: dict[str, Any] = {
        "limit": limit, "offset": offset,
        "include_deleted": include_deleted, "red_team_only": red_team_only,
    }
    for k, v in (("status", status), ("hide_statuses", hide_statuses),
                 ("cve", cve), ("q", q), ("since", since)):
        if v is not None:
            params[k] = v
    if kev_listed is not None:
        params["kev_listed"] = kev_listed
    if rce is not None:
        params["rce"] = rce
    async with _client() as c:
        resp = await c.get(_u("/news/items"), params=params, headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/items/{item_id}")
async def news_item_detail(item_id: str):
    async with _client() as c:
        resp = await c.get(_u(f"/news/items/{item_id}"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.patch("/api/news/items/{item_id}")
async def news_item_patch(item_id: str, body: NewsItemPatch):
    async with _client() as c:
        resp = await c.patch(_u(f"/news/items/{item_id}"),
                             json=body.model_dump(exclude_none=True), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/items/bulk")
async def news_items_bulk(body: NewsBulkBody):
    async with _client() as c:
        resp = await c.post(_u("/news/items/bulk"),
                            json=body.model_dump(), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/items/{item_id}/match-assets")
async def news_item_match_assets(item_id: str):
    async with _client() as c:
        resp = await c.post(_u(f"/news/items/{item_id}/match-assets"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/items/{item_id}/github-search")
async def news_item_github_search(item_id: str):
    async with _client() as c:
        resp = await c.post(_u(f"/news/items/{item_id}/github-search"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/items/{item_id}/enrich")
async def news_item_enrich(item_id: str):
    async with _client() as c:
        resp = await c.post(_u(f"/news/items/{item_id}/enrich"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/deep-search")
async def news_deep_search(body: NewsDeepSearchBody):
    async with _client() as c:
        resp = await c.post(_u("/news/deep-search"),
                            json=body.model_dump(), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/sources")
async def news_sources_list():
    async with _client() as c:
        resp = await c.get(_u("/news/sources"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.patch("/api/news/sources/{source_id}")
async def news_source_patch(source_id: str, body: NewsSourcePatch):
    async with _client() as c:
        resp = await c.patch(_u(f"/news/sources/{source_id}"),
                             json=body.model_dump(exclude_none=True), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.post("/api/news/sources/{source_id}/refetch")
async def news_source_refetch(source_id: str):
    async with _client() as c:
        resp = await c.post(_u(f"/news/sources/{source_id}/refetch"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)


@router.get("/api/news/stats")
async def news_stats():
    async with _client() as c:
        resp = await c.get(_u("/news/stats"), headers=_h())
        if resp.status_code >= 400:
            raise HTTPException(resp.status_code, resp.text)
        return safe_json(resp)
